src/lib/Imp.hs

```diff
@@ -127,15 +148,30 @@ translateDecl env (maybeDest, (Let _ b bound)) = do
 translateExpr :: SubstEnv -> WithDest Expr -> ImpM Atom
 translateExpr env (maybeDest, expr) = case expr of
   Hof hof     -> toImpHof env (maybeDest, hof)
-  App x' idx' -> case getType x' of
-    TabTy _ _ -> do
-      x <- impSubst env x'
-      idx <- impSubst env idx'
-      case x of
-        Lam a@(Abs _ (TabArrow, _)) ->
-          translateBlock mempty (maybeDest, snd $ applyAbs a idx)
-        _ -> error $ "Invalid Imp atom: " ++ pprint x
-    _ -> error $ "shouldn't have non-table app left"
+  App f' x' -> do
+    f <- impSubst env f'
+    x <- impSubst env x'
+    case getType f of
+      TabTy _ _ -> do
+        case f of
+          Lam a@(Abs _ (TabArrow, _)) ->
+            translateBlock mempty (maybeDest, snd $ applyAbs a x)
+          _ -> error $ "Invalid Imp atom: " ++ pprint f
+      _ -> do
+        let (Var (fname:>(FunTy argBinder _ outTy))) = f
+        let argTy = binderAnn argBinder
+        (argDest, argDestElts) <- makeAllocDestWithPtrs Managed $ argTy
+        -- There's a missed opportunity here to use maybeDest. The problem is
+        -- that we can't guarantee that the table refs are in a canonical
+        -- representation. We could consider tracking that information with
+        -- dests so that we can take advantage of it when it happens.
+        (outDest, outDestElts) <- makeAllocDestWithPtrs Managed $ outTy
+        let allElts = outDestElts ++ argDestElts
+        let iFunTy = IFunType CEntryFun (map getIType allElts) []
+        copyAtom argDest x
+        void $ emitStatement $ ICall (fname:>iFunTy) allElts
+        result <- destToAtom outDest
+        copyDest maybeDest result
```

Comment:
This is very confusing. Why do we end up creating the `IFunVar`s here and not in `toImpStandalone`? This feels like it should be just a matter of a simple lookup to me

Reply:
We create an `IFunVar` in both places. We actually don't carry the standalone imp functions in the environment because you can create the call sites without them. You just need the name and the type. In principle we could even compile the caller and callee in parallel.